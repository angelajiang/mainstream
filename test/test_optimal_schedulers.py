import os
import shutil
import subprocess
import pytest

@pytest.mark.unit
@pytest.mark.parametrize("scheduler_type", ["hifi", "stems"])
def test_optimality(scheduler_type):
    output_dir = os.path.join("test", "tmp", scheduler_type)
    if not os.path.exists(output_dir):
        os.makedirs(output_dir)

    schedules_dir = os.path.join(output_dir, "schedules")
    if not os.path.exists(schedules_dir):
        os.makedirs(schedules_dir)
    other_file = os.path.join(schedules_dir, scheduler_type + ".sim.debug.v0")
    exhaustive_file = os.path.join(schedules_dir, "exhaustive.sim.debug.v0")

    subprocess.check_call("test/exhaustive_vs_x_scheduler.sh " + scheduler_type, shell=True)

    other_f1s = []
    with open(other_file, "r") as f:
        for line in f:
            vals = line.split(',')
            metric = round(float(vals[1]), 2)
            other_f1s.append(metric)

    exhaustive_f1s = []
    with open(exhaustive_file, "r") as f:
        for line in f:
            vals = line.split(',')
            metric = round(float(vals[1]), 2)
            exhaustive_f1s.append(metric)

    assert other_f1s == exhaustive_f1s

    shutil.rmtree(output_dir)


@pytest.mark.unit
@pytest.mark.parametrize("scheduler_type,num_apps_range", [
    ("hifi", 5),
    ("stems", 5),
    ("exhaustive", 3),
    ("stems_cpp", 5),
])
def test_regression(regtest, scheduler_type, num_apps_range):
    output_dir = os.path.join("test", "tmp", scheduler_type)
    if not os.path.exists(output_dir):
        os.makedirs(output_dir)

    schedules_dir = os.path.join(output_dir, "schedules")
    if not os.path.exists(schedules_dir):
        os.makedirs(schedules_dir)
    other_file = os.path.join(schedules_dir, scheduler_type + ".sim.debug.v0")

    subprocess.check_call("test/run_scheduler_with_setups.sh {} {}".format(scheduler_type, num_apps_range), shell=True)

    with open(other_file) as f:
        for line in f:
            line = line.split(",")[:-1]
            regtest.write(",".join(line) + "\n")

    shutil.rmtree(output_dir)
