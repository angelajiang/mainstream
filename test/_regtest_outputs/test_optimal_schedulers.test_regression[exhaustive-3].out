d93df00f.v1,2,0.0692267,27,27,4,4,100
7486119a.v1,2,0.20949,27,27,3,4,100
ee938277.v1,2,0.216623,33,3,2,3,100
b3c1e1b9.v1,2,0.282414,33,33,4,3,100
a3956c45.v1,2,0.188819,27,27,4,4,100
0224aea8.v1,3,0.31445,57,69,33,4,4,4,100
2cd3f9a7.v1,3,0.121759,75,27,75,3,4,4,100
32ea467f.v1,3,0.191791,33,69,39,3,4,4,100
2a25c881.v1,3,0.106573,39,33,69,4,3,4,100
42034f03.v1,3,0.31303,39,39,33,4,4,3,100
